### PR TITLE
roleClient

### PR DESCRIPTION
**Pull request by hubjac1**
Implement roleClient and
Split ozone-typescript-client index into files

### CODE REVIEW

File: packages/ozone-typescript-client/src/index.ts [hunk 1]
```diff
@@ -3,10 +3,22 @@ import * as log4javascript from 'log4javascript'
...
+import * as permissionClient from './permissionClient/permissionClient'
+import { PermissionClientImpl } from './permissionClient/permissionClientImpl'
+import * as typeClient from './typeClient/typeClient'
+import { TypeClientImpl } from './typeClient/typeClientImpl'
+
 export namespace OzoneClient {
 	import AssumeStateIsNot = fsm.AssumeStateIsNot
-	import Transitions = fsm.Transitions
-
 	import State = fsm.State
```
Comment (antoinebaudoux) [inline packages/ozone-typescript-client/src/index.ts]: Unused import

File: packages/ozone-typescript-client/src/index.ts [hunk 2]
```diff
@@ -21,207 +33,27 @@ export namespace OzoneClient {
 	import InstalledFilter = httpclient.InstalledFilter
 	import FilterCollection = httpclient.FilterCollection
 
+
 	const log = log4javascript.getLogger('ozone.client')
 	const DEFAULT_TIMEOUT = 5000
```
Comment (antoinebaudoux) [inline packages/ozone-typescript-client/src/index.ts]: Move that in ozoneClientImpl

File: packages/ozone-typescript-client/src/index.ts [hunk 3]
```diff
@@ -21,207 +33,27 @@ export namespace OzoneClient {
...
-		count(query?: Query): Promise<number>
+	export import RoleClient = roleClient.RoleClient
 
-		deleteById(id: UUID, permanent?: boolean): Promise<UUID | null>
-
-		deleteByIds(ids: UUID[], permanent?: boolean): Promise<UUID[]>
-	}
+	export import TypeClient = typeClient.TypeClient
 
 	interface OzoneClientInternals extends OzoneClient {
```
Comment (antoinebaudoux) [inline packages/ozone-typescript-client/src/index.ts]: Move that in ozoneClientImpl

File: packages/ozone-typescript-client/src/index.ts [hunk 4]
```diff
@@ -237,11 +69,7 @@ export namespace OzoneClient {
 
 	const MAX_SESSION_CHECK_DELAY: number = 60000
```
Comment (antoinebaudoux) [inline packages/ozone-typescript-client/src/index.ts]: Move all those constants in ozoneClientImpl

File: packages/ozone-typescript-client/src/index.ts [hunk 5]
```diff
@@ -237,11 +69,7 @@ export namespace OzoneClient {
...
 
-	export interface AuthInfo {
-		principalClass: string,
-		principalId: string,
-		sessionId: string,
-	}
+	export interface AuthInfo extends Credencial.AuthInfo {}
 
 	class Listener {
```
Comment (antoinebaudoux) [inline packages/ozone-typescript-client/src/index.ts]: Move that in ozoneClient

File: packages/ozone-typescript-client/src/index.ts [hunk 6]
```diff
@@ -875,9 +628,7 @@ export namespace OzoneClient {
 		call.headers[name] = value
 	}
 
-	export abstract class OzoneCredentials {
-		abstract authenticate(ozoneURL: string): Promise<AuthInfo>
-	}
+	export abstract class OzoneCredentials extends Credencial.OzoneCredentials {}
 
 	export class UserCredentials extends OzoneCredentials {
```
Comment (antoinebaudoux) [inline packages/ozone-typescript-client/src/index.ts]: Move that in credentials

File: packages/ozone-typescript-client/src/index.ts [hunk 7]
```diff
@@ -237,11 +69,7 @@ export namespace OzoneClient {
...
 
-	export interface AuthInfo {
-		principalClass: string,
-		principalId: string,
-		sessionId: string,
-	}
+	export interface AuthInfo extends Credencial.AuthInfo {}
 
 	class Listener {
 		active: boolean = true
```
Comment (antoinebaudoux) [inline packages/ozone-typescript-client/src/index.ts]: Move OzoneClientImpl in ozoneClientImpl

File: packages/ozone-typescript-client/src/index.ts [hunk 8]
```diff
@@ -950,14 +701,7 @@ export namespace OzoneClient {
...
 
-	export interface ClientConfiguration {
-		ozoneURL: string
-		ozoneInstanceId?: string
-		ozoneCredentials?: OzoneCredentials
-		webSocketsURL?: string
-		defaultTimeout?: number
-	}
-
+	export interface ClientConfiguration extends clientConfiguration.ClientConfiguration {}
```
Comment (antoinebaudoux) [inline packages/ozone-typescript-client/src/index.ts]: Why do you have to do this? Why not simply export the original interface?

Comment (hubjac1) [reply packages/ozone-typescript-client/src/index.ts]: I'don't need I should only do 
export import ClientConfiguration = clientConfiguration.ClientConfiguration

File: packages/ozone-typescript-client/src/typeClient/typeClient.ts [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+	 * delete a type
+	 * @param identifier
+	 */
+	delete(identifier: string): Promise<UUID | null>
+
+	/**
+	 * get list of fields from a type.
+	 * @return {Promise<Array<FieldDescriptor>>} list of field
+	 */
+	getFields(identifier: string): Promise<Array<FieldDescriptor>>
```
Comment (antoinebaudoux) [inline packages/ozone-typescript-client/src/typeClient/typeClient.ts]: This is too specific and has no added value, should not be part of the TypeClient Interface

File: packages/ozone-typescript-client/src/typeClient/typeClient.ts [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+	/**
+	 * get list of all fields (including its parents) from a type.
+	 * TODO should we also get fields from trai?
+	 * @param identifier
+	 */
+	getAllFields(identifier: string): Promise<Array<FieldDescriptor>>
```
Comment (antoinebaudoux) [inline packages/ozone-typescript-client/src/typeClient/typeClient.ts]: This should be part of a client-side type cache system. See the TypeCache class in ozone-client

File: packages/ozone-typescript-client/src/typeClient/typeClient.ts [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+
+	/**
+	 * get list of fields from a type.
+	 * @return {Promise<Array<FieldDescriptor>>} list of field
+	 */
+	getFields(identifier: string): Promise<Array<FieldDescriptor>>
+
+	/**
+	 * get list of all fields (including its parents) from a type.
+	 * TODO should we also get fields from trai?
```
Comment (antoinebaudoux) [inline packages/ozone-typescript-client/src/typeClient/typeClient.ts]: Typo, and yes, you should also get fields from traits

File: packages/ozone-typescript-client/src/typeClient/typeClient.ts [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+	 * @param identifier
+	 */
+	getAllFields(identifier: string): Promise<Array<FieldDescriptor>>
+
+	/**
+	 * verify if the is an instance of an other type
+	 * @param identifier
+	 * @param instance
+	 */
+	isTypeInstanceOf(identifier: string, instance: string): Promise<boolean>
```
Comment (antoinebaudoux) [inline packages/ozone-typescript-client/src/typeClient/typeClient.ts]: This should be part of a client-side type cache system. See the TypeCache class in ozone-client

File: packages/ozone-typescript-client/src/typeClient/typeClientImpl.ts [hunk 13]
```diff
@@ -0,0 +1,120 @@
+import { FromOzone, TypeDescriptor, FieldDescriptor, UUID } from 'ozone-type'
```
Comment (antoinebaudoux) [inline packages/ozone-typescript-client/src/typeClient/typeClientImpl.ts]: Remove unused imports

File: packages/ozone-typescript-client/src/typeClient/typeClientImpl.ts [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+import { TypeClient } from './typeClient'
+import { OzoneClient } from '../ozoneClient/ozoneClient'
+import { httpclient } from 'typescript-http-client'
+import Response = httpclient.Response
```
Comment (antoinebaudoux) [inline packages/ozone-typescript-client/src/typeClient/typeClientImpl.ts]: Remove unused imports

File: packages/ozone-typescript-client/src/typeClient/typeClientImpl.ts [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+import Request = httpclient.Request
+
+export type TypeDescriptorCollection = Map<string, Promise<TypeDescriptor>>
+
+export class TypeClientImpl implements TypeClient {
+
+	constructor(private client: OzoneClient, private baseUrl: string) {
+	}
+
+	static typeCached: TypeDescriptorCollection = new Map<string, Promise<TypeDescriptor>>()
```
Comment (antoinebaudoux) [inline packages/ozone-typescript-client/src/typeClient/typeClientImpl.ts]: This should be part of a Class on top of TypeClient

File: packages/ozone-typescript-client/src/typeClient/typeClientImpl.ts [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+	/**
+	 * Update or create a new type
+	 * @param type
+	 */
+	save(type: TypeDescriptor): Promise<TypeDescriptor> {
+		const request = new Request(`${this.baseUrl}/rest/v3/type`)
+			.setMethod('POST')
+			.setBody(type)
+		const typeDescriptor = this.client.call<TypeDescriptor>(request)
+		TypeClientImpl.typeCached.set(type.identifier, typeDescriptor)
```
Comment (antoinebaudoux) [inline packages/ozone-typescript-client/src/typeClient/typeClientImpl.ts]: This should be part of a Class on top of TypeClient

**Comment by antoinebaudoux** [comment]
> I'm not sure to understand what you mean this "This should be part of a Class on top of TypeClient"

TypeClient should remain barebone and not include caching. It is a good idea to introduce a client-side type cache, but it should sit in a separate Class, and use TypeClient. Also, you should add a cache invalidation mechanism

File: packages/ozone-typescript-client/src/ozoneCredentials/ozoneCredentials.ts [hunk 17]
```diff
@@ -0,0 +1,11 @@
+import { httpclient } from 'typescript-http-client'
+
+export interface AuthInfo {
```
Comment (antoinebaudoux) [inline packages/ozone-typescript-client/src/ozoneCredentials/ozoneCredentials.ts]: AuthInfo should be in "ozoneClient.ts", not "Credentials.ts"

**Pull request merged**